src/parseData.js

```diff
@@ -1,10 +1,15 @@
 /**
  * The resulting parse tree nodes of the parse tree.
  */
-function ParseNode(type, value, mode) {
+function ParseNode(type, value, mode, firstToken, lastToken) {
     this.type = type;
     this.value = value;
     this.mode = mode;
+    if (firstToken) {
+        this.lexer = firstToken.lexer;
+        this.start = firstToken.start;
+        this.end = (lastToken || firstToken).end;
+    }
```

Comment:
This is all about error reporting. Ideally, we should be able to take an arbitrary piece of parse tree and highlight its corresponding source region. For this we need the start and end positions as well as the input string. And the input string is stored in the lexer, thus we need the lexer.

We might conceivably use just the input string instead of the lexer, both here and for `Token`. But using the lexer allows us to add metadata to a source string at some later point. Things like file name, or some other description of where this string came from. For example “in definition of macro …” for source regions coming from an external macro definition string would already make a lot of sense, and be fairly easy with the current approach.